File: projetos/jogo-da-velha/jogo_da_velha.py
import random

def print_board(board):
    for row in board:
        print(" | ".join(row))
        print("-" * 5)

def check_winner(board, player):
    
    for i in range(3):
        if all([cell == player for cell in board[i]]): 
            return True
        if all([board[j][i] == player for j in range(3)]):  
            return True
    if all([board[i][i] == player for i in range(3)]):  
        return True
    if all([board[i][2-i] == player for i in range(3)]):  
        return True
    return False


def is_full(board):
    return all([cell != ' ' for row in board for cell in row])

def minimax(board, depth, is_maximizing, alpha, beta):
    if check_winner(board, 'X'):
        return -1  
    if check_winner(board, 'O'):
        return 1  
    if is_full(board):
        return 0  
    if is_maximizing:
        max_eval = -float('inf')
        for i in range(3):
            for j in range(3):
                if board[i][j] == ' ':
                    board[i][j] = 'O'
                    eval = minimax(board, depth + 1, False, alpha, beta)
                    board[i][j] = ' '
                    max_eval = max(max_eval, eval)
                    alpha = max(alpha, eval)
                    if beta <= alpha:
                        break
        return max_eval
    else:
        min_eval = float('inf')
        for i in range(3):
            for j in range(3):
                if board[i][j] == ' ':
                    board[i][j] = 'X'
                    eval = minimax(board, depth + 1, True, alpha, beta)
                    board[i][j] = ' '
                    min_eval = min(min_eval, eval)
                    beta = min(beta, eval)
                    if beta <= alpha:
                        break
        return min_eval


def best_move(board):
    max_eval = -float('inf')
    move = None
    for i in range(3):
        for j in range(3):
            if board[i][j] == ' ':
                board[i][j] = 'O'
                eval = minimax(board, 0, False, -float('inf'), float('inf'))
                board[i][j] = ' '
                if eval > max_eval:
                    max_eval = eval
                    move = (i, j)
    return move

def play_game():
    board = [[' ' for _ in range(3)] for _ in range(3)]
    while True:
        print_board(board)
        
        row, col = map(int, input("Digite a linha e a coluna (0-2) para sua jogada: ").split())
        if board[row][col] != ' ':
            print("Movimento inválido. Tente novamente.")
            continue
        board[row][col] = 'X'
        
        if check_winner(board, 'X'):
            print_board(board)
            print("Você venceu!")
            break
        if is_full(board):
            print_board(board)
            print("Empate!")
            break
        
        print("A IA está jogando...")
        move = best_move(board)
        board[move[0]][move[1]] = 'O'
        
        if check_winner(board, 'O'):
            print_board(board)
            print("A IA venceu!")
            break
        if is_full(board):
            print_board(board)
            print("Empate!")
            break


play_game()

File: projetos/jogo-da-velha/jogo_da_velha.c
#include <stdio.h>

char tabuleiro[9] = {'1','2','3','4','5','6','7','8','9'};

void exibir_tabuleiro() {
    printf("\n %c | %c | %c\n", tabuleiro[0], tabuleiro[1], tabuleiro[2]);
    printf("---+---+---\n");
    printf(" %c | %c | %c\n", tabuleiro[3], tabuleiro[4], tabuleiro[5]);
    printf("---+---+---\n");
    printf(" %c | %c | %c\n\n", tabuleiro[6], tabuleiro[7], tabuleiro[8]);
}

int verificar_vitoria() {
    int v[8][3] = {
        {0,1,2},{3,4,5},{6,7,8},
        {0,3,6},{1,4,7},{2,5,8},
        {0,4,8},{2,4,6}
    };
    for (int i = 0; i < 8; i++) {
        if (tabuleiro[v[i][0]] == tabuleiro[v[i][1]] && tabuleiro[v[i][1]] == tabuleiro[v[i][2]])
            return 1;
    }
    return 0;
}

int main() {
    int jogadas = 0;
    char jogador = 'X';

    while (1) {
        exibir_tabuleiro();
        int pos;
        printf("Jogador %c, escolha uma posição (1-9): ", jogador);
        scanf("%d", &pos);
        pos--;

        if (pos < 0 || pos > 8 || tabuleiro[pos] == 'X' || tabuleiro[pos] == 'O') {
            printf("Jogada inválida. Tente novamente.\n");
            continue;
        }

        tabuleiro[pos] = jogador;
        jogadas++;

        if (verificar_vitoria()) {
            exibir_tabuleiro();
            printf("Jogador %c venceu!\n", jogador);
            break;
        }

        if (jogadas == 9) {
            exibir_tabuleiro();
            printf("Empate!\n");
            break;
        }

        jogador = (jogador == 'X') ? 'O' : 'X';
    }

    return 0;
}
